CommunityToolkit.Mvvm/Input/AsyncRelayCommand.cs

```diff
@@ -97,27 +147,65 @@ public AsyncRelayCommand(Func<CancellationToken, Task> cancelableExecute, Func<b
     {
         this.cancelableExecute = cancelableExecute;
         this.canExecute = canExecute;
+        this.allowConcurrentExecutions = true;
     }
 
-    private TaskNotifier? executionTask;
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AsyncRelayCommand"/> class.
+    /// </summary>
+    /// <param name="cancelableExecute">The cancelable execution logic.</param>
+    /// <param name="canExecute">The execution status logic.</param>
+    /// <param name="allowConcurrentExecutions">Whether or not to allow concurrent executions of the command.</param>
+    public AsyncRelayCommand(Func<CancellationToken, Task> cancelableExecute, Func<bool> canExecute, bool allowConcurrentExecutions)
+    {
+        this.cancelableExecute = cancelableExecute;
+        this.canExecute = canExecute;
+        this.allowConcurrentExecutions = allowConcurrentExecutions;
+    }
+
+    private Task? executionTask;
 
     /// <inheritdoc/>
     public Task? ExecutionTask
     {
         get => this.executionTask;
         private set
         {
-            if (SetPropertyAndNotifyOnCompletion(ref this.executionTask, value, _ =>
+            if (ReferenceEquals(this.executionTask, value))
             {
-                    // When the task completes
-                    OnPropertyChanged(IsRunningChangedEventArgs);
-                OnPropertyChanged(CanBeCanceledChangedEventArgs);
-            }))
+                return;
+            }
+
+            this.executionTask = value;
+
+            PropertyChanged?.Invoke(this, ExecutionTaskChangedEventArgs);
+            PropertyChanged?.Invoke(this, IsRunningChangedEventArgs);
+            PropertyChanged?.Invoke(this, CanBeCanceledChangedEventArgs);
+
+            if (value?.IsCompleted ?? true)
             {
-                // When setting the task
-                OnPropertyChanged(IsRunningChangedEventArgs);
-                OnPropertyChanged(CanBeCanceledChangedEventArgs);
+                return;
             }
+
+            static async void MonitorTask(AsyncRelayCommand @this, Task task)
+            {
+                try
+                {
+                    await task;
+                }
+                catch
+                {
+                }
+
+                if (ReferenceEquals(@this.executionTask, task))
+                {
+                    @this.PropertyChanged?.Invoke(@this, ExecutionTaskChangedEventArgs);
+                    @this.PropertyChanged?.Invoke(@this, IsRunningChangedEventArgs);
+                    @this.PropertyChanged?.Invoke(@this, CanBeCanceledChangedEventArgs);
+                }
+            }
+
+            MonitorTask(this, value!);
```

Comment:
That is correct: this is to guard against potential concurrent modifications. In thery that's not really necessary as you should always be using this command from the UI thread (meaning you're implicitly synchronized already), but I figured (1) better safe than sorry anyway and (2) using `value` here is faster anyway given the value is already there, instead of having to indirectly load it from a field. Functionally it's basically the same though, yeah 😄